tests/contrib/funsor/test_vectorized_markov.py

```diff
@@ -0,0 +1,388 @@
+    w_trans = pyro.param("w_trans", lambda: torch.rand((w_dim, w_dim)), constraint=constraints.simplex)
+    x_init = pyro.param("x_init", lambda: torch.rand(x_dim), constraint=constraints.simplex)
+    x_trans = pyro.param("x_trans", lambda: torch.rand((x_dim, x_dim)), constraint=constraints.simplex)
+    y_probs = pyro.param("y_probs", lambda: torch.rand(w_dim, x_dim, y_dim), constraint=constraints.simplex)
+
+    w_prev = x_prev = None
+    markov_loop = \
+        pyro.vectorized_markov(name="time", size=len(data), dim=-2, history=history) if vectorized \
+        else pyro.markov(range(len(data)), history=history)
+    for i in markov_loop:
+        w_curr = pyro.sample(
+            "w_{}".format(i), dist.Categorical(
+                w_init if isinstance(i, int) and i < 1 else w_trans[w_prev]),
+            infer={"enumerate": "parallel"})
+        x_curr = pyro.sample(
+            "x_{}".format(i), dist.Categorical(
+                x_init if isinstance(i, int) and i < 1 else x_trans[x_prev]),
+            infer={"enumerate": "parallel"})
+        with pyro.plate("tones", data.shape[-1], dim=-1):
+            pyro.sample("y_{}".format(i), dist.Categorical(
+                Vindex(y_probs)[w_curr, x_curr]),
+                obs=data[i])
+        x_prev, w_prev = x_curr, w_curr
+
+
+def model_4(data, history, vectorized):
+    w_dim, x_dim, y_dim = 2, 3, 2
+    pyro.set_rng_seed(0)
+    pyro.get_param_store().clear()
+    w_init = pyro.param("w_init", lambda: torch.rand(w_dim), constraint=constraints.simplex)
+    w_trans = pyro.param("w_trans", lambda: torch.rand((w_dim, w_dim)), constraint=constraints.simplex)
+    x_init = pyro.param("x_init", lambda: torch.rand(w_dim, x_dim), constraint=constraints.simplex)
+    x_trans = pyro.param("x_trans", lambda: torch.rand((w_dim, x_dim, x_dim)), constraint=constraints.simplex)
+    y_probs = pyro.param("y_probs", lambda: torch.rand(w_dim, x_dim, y_dim), constraint=constraints.simplex)
+
+    w_prev = x_prev = None
+    markov_loop = \
+        pyro.vectorized_markov(name="time", size=len(data), dim=-2, history=history) if vectorized \
+        else pyro.markov(range(len(data)), history=history)
+    for i in markov_loop:
+        w_curr = pyro.sample(
+            "w_{}".format(i), dist.Categorical(
+                w_init if isinstance(i, int) and i < 1 else w_trans[w_prev]),
+            infer={"enumerate": "parallel"})
+        x_curr = pyro.sample(
+            "x_{}".format(i), dist.Categorical(
+                x_init[w_curr] if isinstance(i, int) and i < 1 else x_trans[w_curr, x_prev]),
+            infer={"enumerate": "parallel"})
+        with pyro.plate("tones", data.shape[-1], dim=-1):
+            pyro.sample("y_{}".format(i), dist.Categorical(
+                Vindex(y_probs)[w_curr, x_curr]),
+                obs=data[i])
+        x_prev, w_prev = x_curr, w_curr
+
+
+def model_5(data, history, vectorized):
```

Comment:
Awesome! IIUC, this api supports history>=2. Is there any requirement/restriction for the code to achieve that?

@eb8680 Would this be a solution for https://github.com/pyro-ppl/numpyro/issues/702? Which API should I use for the future: `scan` or `vectorized_markov`? :D

Reply:
Yup, it does! Here is an example written using both `pyro.markov` and `pyro.vectorized_markov`:

```py
def model_5(data, history=2, vectorized=True):
    x_dim, y_dim = 3, 2
    pyro.set_rng_seed(0)
    pyro.get_param_store().clear()
    x_init = pyro.param("x_init", lambda: torch.rand(x_dim), constraint=constraints.simplex)
    x_init_2 = pyro.param("x_init_2", lambda: torch.rand(x_dim, x_dim), constraint=constraints.simplex)
    x_trans = pyro.param("x_trans", lambda: torch.rand((x_dim, x_dim, x_dim)), constraint=constraints.simplex)
    y_probs = pyro.param("y_probs", lambda: torch.rand(x_dim, y_dim), constraint=constraints.simplex)

    x_prev = x_prev_2 = None
    markov_loop = \
        pyro.vectorized_markov(name="time", size=len(data), dim=-2, history=history) if vectorized \
        else pyro.markov(range(len(data)), history=history)
    for i in markov_loop:
        if isinstance(i, int) and i == 0:
            x_probs = x_init
        elif isinstance(i, int) and i == 1:
            x_probs = Vindex(x_init_2)[x_prev]
        else:
            x_probs = Vindex(x_trans)[x_prev_2, x_prev]

        x_curr = pyro.sample(
            "x_{}".format(i), dist.Categorical(x_probs),
            infer={"enumerate": "parallel"})
        with pyro.plate("tones", data.shape[-1], dim=-1):
            pyro.sample("y_{}".format(i), dist.Categorical(
                Vindex(y_probs)[x_curr]),
                obs=data[i])
        x_prev_2, x_prev = x_prev, x_curr
```

Reply:
Sorry, I completely missed the reference to the code!

Reply:
My hope is that any code that can be written with `pyro.markov` (using proper `init` probabilities) can be replaced with `pyro.vectorized_markov`.

Reply:
> IIUC, this api supports history>=2. Is there any requirement/restriction for the code to achieve that?

Well, right now the corresponding tensor variable elimination algorithm for `history > 1` doesn't exist - the one in https://github.com/pyro-ppl/funsor/pull/398 only supports `history <= 1`.  The enumeration behavior in this PR should be general, though.

> Would this be a solution for pyro-ppl/numpyro#702? Which API should I use for the future: scan or vectorized_markov? :D

Once this code is finished, I think it might make sense to write an equivalent JAX-flavored version in terms of `scan` for NumPyro, which would be a realization of https://github.com/pyro-ppl/numpyro/issues/686.  That should be pretty easy after the details are worked out here. As @ordabayevy says, our design goal is to make `pyro.vectorized_markov` interchangeable with `pyro.markov` whenever that is mathematically permissible.

Reply:
I see, thanks for explaining! This API should be convenient for many usage cases (though I guess `scan` is still useful when it is tricky to write vectorized code) and the code might look more intuitive with `for` loop. In addition, this does not require a scan loop to collect the values, which enables a performance boost.